### ISSUE TITLE
grade calculation for multi grade/subject curriculum items

### ISSUE THREAD

**Issue by endoooo** [camino-school/lanttern#109]
<img width="475" alt="image" src="https://github.com/camino-school/lanttern/assets/746889/722e7167-9f57-4f03-b962-454f3092a794">

assessment points out of the grade composition scope are being considered in the calculation.

my hypothesis in that there's some join in the query related to the curriculum item that is causing this.

**Comment by endoooo** [comment]
actually, I don't think it's happening — it was caused by a misplaced assessment point entry registration.
I'll close it for now.

**Comment by endoooo** [comment]
evidences from https://lanttern.org/report_cards/3?tab=grades&is_editing_grade_composition=17 (grade 3 report card)

<img width="541" alt="image" src="https://github.com/camino-school/lanttern/assets/746889/db40fb7d-a8d2-4283-9137-f0f8e8b7de03">
movement composition

<img width="518" alt="image" src="https://github.com/camino-school/lanttern/assets/746889/fd7c8098-5d5c-4578-9207-ade2ca9e09f5">
example of a student's movement grade composition

<img width="965" alt="Screenshot 2024-04-11 at 13 25 05" src="https://github.com/camino-school/lanttern/assets/746889/a31a5013-6b5b-464e-82b2-a5fb2c9dedfa">
wrong marking in grade 2 expedition

---

although in the grade composition we pick only the grade 3 expedition, the calculation is also considering the wrong marking in grade 2 expedition. I think it's happening because it's the same curriculum item — it's a good starting point

**Comment by endoooo** [comment]
### new findings (context: expressions G2 and G3)

if you go to G2 report card and add the G3 expedition (or vice-versa), the G3 curriculum items that should be part only of the G3 grades are shown in the G2 grade composition. if you try to remove them from the G2 composition, it affects the G3 composition as well

**Comment by endoooo** [comment]
### "bug" discovered. info below

this issue exists as the result of a bad design decision. currently, we have the same `GradesReport` linked to multiple `ReportCard`s, from different years. it kinda works because we expect that, for each `GradesReport` sub cycle (e.g. 2024 Q1), the students will be graded based on assessments that happen in the context of strands related to the students current school year.

the problem with this design is that there is no isolation between years, and when teachers assess students (by mistake) in the wrong strand/year, this assessment impacts the whole grading composition because all `GradeComponents`s are linked to a `ReportCard`, which is linked to a `GradesReport`, which is the same used in many other `ReportCard`s.

there are two possible initial solutions for this issue:

1. the first and easier one is to add the report card id as an additional (and required) argument for `calculate_student_grade/4`, `calculate_student_grades/3`, `calculate_subject_grades/3`, and `calculate_cycle_grades/3`, so we can calculate the grades considering only the grades components that are linked to the given report card.

2. re-structure the whole grades report architecture, maybe adding a mandatory `year_id` field to `GradesReport` and linking the `GradeComponent`s to the grade report.

we'll go with option 2, to isolate the grades reports from the report cards — which is an important architectural decision for future features, like importing past (pre-Lanttern) students grades, or viewing the students' grades history.

in short, we need to remove the current interdependence between grades reports and report cards (mainly due to grades composition components being linked to report cards), an move towards a structure where they are both independent entities — grades reports can be viewed and composed in the context of a report card, but it should not be dependent of the report card.
